sky/skylet/providers/ibm/node_provider.py

```diff
@@ -0,0 +1,924 @@
+                node["id"], nic_id
+            ).get_result()
+
+            floating_ips = res["floating_ips"]
+            if len(floating_ips) == 0:
+                # not adding a node that's missing floating ip
+                continue
+            else:
+                node["floating_ips"] = floating_ips
+
+            res_nodes.append(node)
+
+        for node in res_nodes:
+            self.cached_nodes[node["id"]] = node
+
+        return [node["id"] for node in res_nodes]
+
+    def is_running(self, node_id) -> bool:
+        """returns whether a node is in status running"""
+        with self.lock:
+            node = self._get_cached_node(node_id)
+            logger.debug(
+                f"""node: {node_id} is_running?
+                {node["status"] == "running"}""")
+            return node["status"] == "running"
+
+    def is_terminated(self, node_id) -> bool:
+        """returns True if a node is either not recorded or
+         not in any valid status."""
+        with self.lock:
+            try:
+                node = self._get_cached_node(node_id)
+                logger.debug(
+                    f"""node: {node_id} is_terminated?
+                    {node["status"] not in
+                    ["running", "starting", "pending"]}""")
+                return node["status"] not in ["running", "starting", "pending"]
+            # pylint: disable=W0703
+            except Exception:
+                return True
+
+    def node_tags(self, node_id) -> Dict[str, str]:
+        """returns tags of specified node id """
+
+        with self.lock:
+            return self.nodes_tags.get(node_id, {})
+
+    def _get_hybrid_ip(self, node_id):
+        """return external ip for head and private ips for workers"""
+
+        # skypilot-adjustment returns external ips for worker nodes as well.
+        node = self._get_cached_node(node_id)
+        fip = node.get("floating_ips")
+        if fip:
+            return fip[0]["address"]
+
+        node = self._get_node(node_id)
+        fip = node.get("floating_ips")
+        if fip:
+            return fip[0]["address"]
+
+    def external_ip(self, node_id) -> str:
+        """returns head node's public ip.
+        if use_hybrid_ips==true in cluster's config file,
+         returns the ip address of a node based on its 'Kind'."""
+
+        with self.lock:
+            if self.provider_config.get("use_hybrid_ips"):
+                return self._get_hybrid_ip(node_id)
+
+            node = self._get_cached_node(node_id)
+            fip = node.get("floating_ips")
+            if fip:
+                return fip[0]["address"]
+
+    def internal_ip(self, node_id) -> str:
+        """returns the worker's node private ip address"""
+        node = self._get_cached_node(node_id)
+
+        try:
+            primary_ip = node["network_interfaces"][0].get("primary_ip")[
+                "address"]
+            if primary_ip is None:
+                node = self._get_node(node_id)
```

Comment:
Agreed, but it's used only as a safe guard, i.e. it fetches node data from cache and only if the VPC platform fails (which it shouldn't) to update the internal ip field, we'll run a query to update it. It could be a VPC bug (that could've been resolved by now). 
As performance isn't taking a hit (code shouldn't ever be executed), would you be content with me adding documentation?

Reply:
I see. Usually I prefer less redundant code, but it sounds like there might be some nasty VPC bugs lurking around. So I think adding documentation sounds good.

Reply:
Sure, added.